packages/hub/src/utils/createXorbs.ts

```diff
@@ -0,0 +1,121 @@
+/**
+ * Todo: add dedup: we actually need to remember chunks already written, and not add themm to the xorb, and also
+ *  take that into account for file reconstruction
+ * Todo: byte grouping?
+ */
+
+import { XET_CHUNK_HEADER_BYTES, XetChunkCompressionScheme } from "./XetBlob";
+import { compress as lz4_compress } from "../vendor/lz4js";
+
+const TARGET_CHUNK_SIZE = 64 * 1024;
+const XORB_SIZE = 64 * 1024 * 1024;
+const MAX_XORB_CHUNKS = 8 * 1024;
+
+export async function* createXorbs(
+	fileSource: Blob
+): AsyncGenerator<{ xorb: Uint8Array; hash: string }, void, undefined> {
+	const chunkModule = await import("../vendor/xet-chunk/chunker_wasm");
+	await chunkModule.init();
+	const chunker = new chunkModule.Chunker(TARGET_CHUNK_SIZE);
+
+	let xorb = new Uint8Array(XORB_SIZE);
+	const sourceChunks: Array<Uint8Array> = [];
+
+	try {
+		const reader = fileSource.stream().getReader();
+		let xorbOffset = 0;
+		let xorbChunks = Array<{ hash: string; length: number }>();
+
+		const addChunks = function* (chunks: Array<{ hash: string; length: number }>) {
+			for (const chunk of chunks) {
+				let chunkToCopy: Uint8Array;
+				if (chunk.length === sourceChunks[0].length) {
+					chunkToCopy = sourceChunks[0];
+					sourceChunks.shift();
```

Comment:
This looks more inefficient than using an index approach no?

Reply:
this is just the leftover data from the chunking process, there should be at most 128kB of it I think and even in 16kB chunks it's only 8 chunks, eg `sourceChunks.length <= 8`

So with those params I don't think we care too much